File: machine/cpp/rompack/tooling/blua32_media.cpp
#include "rompack/tooling/blua32_media.h"

#include "rompack/image.h"
#include "rompack/toc.h"

#include <utility>

namespace bmsx {

auto loadBlua32ToolingImage(
	const RomImage& rom,
	u32 romBaseAddress
) -> std::optional<Blua32ToolingImage> {
	std::optional<Blua32ImageLayout> layout =
		decodeBlua32RomImage(rom.bytes, romBaseAddress);
	if (!layout) {
		return std::nullopt;
	}
	std::optional<Blua32SymbolsImage> symbols;
	const RomTocPayload toc = decodeRomToc(
		rom.bytes.data() + rom.header.tocOffset,
		rom.header.tocLength
	);
	for (const RomSourceEntry& entry : toc.entries) {
		if (entry.resid == BLUA32_SYMBOLS_IMAGE_ID) {
			const size_t start = static_cast<size_t>(*entry.rom.start);
			const size_t byteCount =
				static_cast<size_t>(*entry.rom.end - *entry.rom.start);
			symbols.emplace(
				decodeBlua32SymbolsImage(rom.bytes.subspan(start, byteCount))
			);
			break;
		}
	}
	return Blua32ToolingImage{
		std::move(*layout),
		std::move(symbols),
	};
}

auto blua32ToolingImageForDomain(
	const Blua32ToolingMedia& media,
	ExecutionDomainId executionDomainId
) -> const Blua32ToolingImage* {
	if (executionDomainId < 0) {
		return media.system ? &*media.system : nullptr;
	}
	const std::optional<Blua32ToolingImage>& image =
		media.cartridgeSlots[static_cast<size_t>(executionDomainId)];
	return image ? &*image : nullptr;
}

} // namespace bmsx

File: machine/ts/rompack/tooling/blua32_media.ts
import {
	decodeBlua32RomImage,
	type Blua32ImageLayout,
} from './blua32_image';
import {
	BLUA32_SYMBOLS_IMAGE_ID,
	decodeBlua32SymbolsImage,
	type Blua32SymbolsImage,
} from './blua32_symbols';
import type { ExecutionDomainId } from '../../spec/blua32/execution_domain';
import type { RomToolingLayer } from './loader';
import type { RawRomSource } from './source';

export type Blua32ToolingImage = {
	readonly layout: Blua32ImageLayout;
	readonly symbols: Blua32SymbolsImage | null;
};

export type Blua32ToolingMedia = {
	readonly system: Blua32ToolingImage | null;
	readonly cartridgeSlots: readonly [
		Blua32ToolingImage | null,
		Blua32ToolingImage | null,
	];
};

export function loadBlua32ToolingImage(
	rom: RomToolingLayer,
	romSource: RawRomSource,
	romBaseAddress: number,
): Blua32ToolingImage | null {
	const layout = decodeBlua32RomImage(rom.payload, romBaseAddress);
	if (!layout) {
		return null;
	}
	const symbolsEntry = romSource.getEntry(BLUA32_SYMBOLS_IMAGE_ID);
	return {
		layout,
		symbols: symbolsEntry
			? decodeBlua32SymbolsImage(romSource.getBytesView(symbolsEntry))
			: null,
	};
}

export function blua32ToolingImageForDomain(
	media: Blua32ToolingMedia,
	executionDomainId: ExecutionDomainId,
): Blua32ToolingImage | null {
	switch (executionDomainId) {
		case -1:
			return media.system;
		case 0:
			return media.cartridgeSlots[0];
		case 1:
			return media.cartridgeSlots[1];
	}
}
